src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs

```diff
@@ -0,0 +1,274 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use anyhow::{anyhow, Context, Result};
+
+use std::{collections::HashMap, sync::Arc};
+
+use agent::Agent;
+use async_trait::async_trait;
+use common::{
+    error::Error,
+    types::{
+        ContainerConfig, ContainerID, ContainerProcess, ExecProcessRequest, KillRequest,
+        ProcessExitStatus, ProcessStateInfo, ProcessType, ResizePTYRequest, ShutdownRequest,
+        StatsInfo, UpdateRequest, PID,
+    },
+    ContainerManager,
+};
+use oci::Process as OCIProcess;
+use resource::ResourceManager;
+use tokio::sync::RwLock;
+
+use super::{logger_with_process, Container};
+
+unsafe impl Send for VirtContainerManager {}
+unsafe impl Sync for VirtContainerManager {}
+pub struct VirtContainerManager {
+    sid: String,
+    pid: u32,
+    containers: Arc<RwLock<HashMap<String, Container>>>,
+    resource_manager: Arc<ResourceManager>,
+    agent: Arc<dyn Agent>,
+}
+
+impl VirtContainerManager {
+    pub fn new(
+        sid: &str,
+        pid: u32,
+        agent: Arc<dyn Agent>,
+        resource_manager: Arc<ResourceManager>,
+    ) -> Self {
+        Self {
+            sid: sid.to_string(),
+            pid,
+            containers: Default::default(),
+            resource_manager,
+            agent,
+        }
+    }
+}
+
+#[async_trait]
+impl ContainerManager for VirtContainerManager {
+    async fn create_container(&self, config: ContainerConfig, spec: oci::Spec) -> Result<PID> {
+        let container = Container::new(
+            self.pid,
+            config,
+            self.agent.clone(),
+            self.resource_manager.clone(),
+        )
+        .context("new container")?;
+
+        let mut containers = self.containers.write().await;
+        container.create(spec).await.context("create")?;
+        containers.insert(container.container_id.to_string(), container);
+
+        Ok(PID { pid: self.pid })
+    }
+
+    async fn close_process_io(&self, process: &ContainerProcess) -> Result<()> {
+        let containers = self.containers.read().await;
+        let container_id = &process.container_id.to_string();
+        let c = containers
+            .get(container_id)
+            .ok_or_else(|| Error::ContainerNotFound(container_id.clone()))?;
+
+        c.close_io(process).await.context("close io")?;
+        Ok(())
+    }
+
+    async fn delete_process(&self, process: &ContainerProcess) -> Result<ProcessStateInfo> {
+        let container_id = &process.container_id.container_id;
+        match process.process_type {
+            ProcessType::Container => {
+                let mut containers = self.containers.write().await;
+                let c = containers
+                    .remove(container_id)
+                    .ok_or_else(|| Error::ContainerNotFound(container_id.to_string()))?;
+                c.state_process(process).await.context("state process")
```

Comment:
What's the `state_process` meaning here?

Reply:
Thanks for your review, `state_process` try to get process state for shim API need.